Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+            longest = &left;
+            NSDecimalCopy(result, &left);
+
+        } else {
+            shortLen = leftOperand->_length;
+            shortest = &left;
+            longestLen = rightOperand->_length;
+            longest = &right;
+            NSDecimalCopy(result, &right);
+        }
+
+        unsigned short finalResult[NSDecimalMaxSize + 1] = { 0 };
+        unsigned int res = 0;
+
+        for (int index = 0; index < shortLen; ++index) {
+            res += shortest->_mantissa[index] + longest->_mantissa[index];
+            finalResult[index] = res;
+            res = res >> 0x10;
+        }
+
+        // take care of the rest
+        for (int index = shortLen; index < longestLen; ++index) {
+            res += longest->_mantissa[index];
+            finalResult[index] = res;
+            res = res >> 0x10;
+        }
+
+        if (res != 0) {
+            if (longestLen >= NSDecimalMaxSize + 1) {
+                INVALIDATE_DECIMAL(result);
+                return NSCalculationOverflow;
+            }
+            finalResult[longestLen] = res;
+            result->_length = longestLen + 1;
+        }
+
+        // We need to fit
+        NSDecimalInternal internalResult = { 0 };
+        NSDecimalCopy(&internalResult.decimal, result);
+        internalResult._mantissa = finalResult;
+        internalResult._remainder = 0;
+        internalResult._extendedLen = NSDecimalMaxSize + 1;
+
+        NSCalculationError error = _compactMantissa(&internalResult, roundingMode);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+
+        // set the result back
+        _InternalDecimalCopy(result, &internalResult);
+        NSDecimalCompact(result);
+        return NSCalculationNoError;
+    }
+
+    return _handleDifferenceOfSignAddition(result, &left, &right, roundingMode);
+}
+
+/**
+ @Status Caveat
+ @Notes roundingMode is not supported.
+*/
+NSCalculationError NSDecimalSubtract(NSDecimal* result,
+                                     const NSDecimal* leftOperand,
+                                     const NSDecimal* rightOperand,
+                                     NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL) && (result != NULL));
+    VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand);
+    // Convert the subtraction into an addition.
+    // L - R -> L + (-R)
+    // flip the sign, if and only if it's non zero and a valid number
+    NSDecimal right;
+    NSDecimalCopy(&right, rightOperand);
+    right._isNegative = (IS_ZERO(&right) || NSDecimalIsNotANumber(&right)) ? right._isNegative : !right._isNegative;
+
+    return NSDecimalAdd(result, leftOperand, &right, roundingMode);
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSCalculationError NSDecimalMultiplyByPowerOf10(NSDecimal* result, const NSDecimal* number, short power, NSRoundingMode roundingMode) {
```

Comment:
Also need to take care of cases for exp + power >= 127 and the lower -127 limit. 
I'll add this as separate CR (given I need to write extensive tests to test edge cases)

---
In reply to: [80589824](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80589824) [](ancestors = 80589824)